### PR TITLE
Enhance search-based code intel with accurate local navigation.

### PR DESCRIPTION
**Pull request by olafurpg**
Previously, search-based code intel did not support navigating to local
definitions because ctags doesn't index local variables. This commit
adds a new `searchBasedLsif()` endpoint to the GraphQL API that
search-based code intel can use to resolve accurate code navigation
for symbols that are definitely local.

- This only works for Go right now but we can add support for more
  languages in the future. We use tree-sitter to parse the code because
  it provides a nice API but we can use other parsers too.
- We use the new "LSIF typed" data schema from RFC 519 to index
  search-based code intel. The benefit of using this schema is
  that we can reuse testing infrastructure and vocabulary between
  search-based and precise indexers.
- This commit adds new snapshot testing infrastructure that is similar
  to how we do snapshot testing in lsif-java. It would be nice to extract
  this logic into a standalone tool so that it can be consumed by
  precise indexers in the future.
- The new `searchBasedLsif` endpoint can be extended in the future
  to implement stencil support for search-based code intel. It's
  also worth exploring whether we should use the same infrastructure to
  implement syntax highlighting.

### CODE REVIEW

**Comment by olafurpg** [comment]
>  error: failed to index: failed to load packages: packages.Load: err: exit status 1: stderr: go build github.com/smacker/go-tree-sitter/golang: build constraints exclude all Go files in /go/pkg/mod/github.com/smacker/go-tree-sitter@v0.0.0-20211116060328-db7fde9b5e82/golang

Any advice on how to troubleshoot this error from the lsif-go CI job?

**Comment by tjdevries** [comment]
This looks super cool :) Want to chat about it some time this week?

File: cmd/frontend/graphqlbackend/schema.graphql [hunk 1]
```diff
@@ -3279,6 +3279,47 @@ type Position {
     character: Int!
 }
 
+"""
```
Comment (efritz) [inline cmd/frontend/graphqlbackend/schema.graphql:3282]: It doesn't make sense to me that any of this is called LSIF.

**Review by efritz — CHANGES_REQUESTED** [review]
General approach looks good to me (modulo naming and graphql API consistency). I have issue using either _LSIF_ or _search-based_ as a term here as both are incorrect. Would be nice to separate the graphql and the search API into a shared package to keep the graphqlbackend layer devoid of any useful logic.

**Comment by olafurpg** [comment]
Closing this PR in favor of https://github.com/sourcegraph/sourcegraph/pull/29427, based on discussions from RFC 555 https://docs.google.com/document/d/1IsC8NydIDQ3e_lEBqB1c084-u6BJc1dw0KmS2N0s2Eo/edit

**Pull request closed without merging**